services/wallet/router/fees/estimated_time.go

```diff
@@ -102,14 +115,167 @@ func (f *FeeManager) estimatedTime(feeHistory *FeeHistory, maxFeePerGas *big.Int
 }
 
-func (f *FeeManager) getFeeHistorySorted(feeHistory *FeeHistory) ([]*big.Int, error) {
-	fees := []*big.Int{}
-	for _, fee := range feeHistory.BaseFeePerGas {
-		i := new(big.Int)
-		i.SetString(strings.Replace(fee, "0x", "", 1), 16)
-		fees = append(fees, i)
+func (f *FeeManager) convertToBigIntAndSort(hexArray []string) []*big.Int {
+	values := []*big.Int{}
+	for _, sValue := range hexArray {
+		iValue := new(big.Int)
+		_, ok := iValue.SetString(sValue, 0)
+		if !ok {
+			continue
+		}
+		values = append(values, iValue)
+	}
+
+	sort.Slice(values, func(i, j int) bool { return values[i].Cmp(values[j]) < 0 })
+	return values
+}
+
+// TransactionEstimatedTimeV2 returns the estimated time in seconds for a transaction to be included in a block
+func (f *FeeManager) TransactionEstimatedTimeV2(ctx context.Context, chainID uint64, maxFeePerGas *big.Int, priorityFee *big.Int) uint {
+	feeHistory, err := f.getFeeHistory(ctx, chainID, 100, "latest", []int{RewardPercentiles2})
+	if err != nil {
+		return 0
+	}
+
+	return f.estimatedTimeV2(feeHistory, maxFeePerGas, priorityFee, chainID)
+}
+
+func calculateTimeForInclusion(chainID uint64, expectedInclusionInBlock int) uint {
+	blockCreationTime := common.GetBlockCreationTimeForChain(chainID)
+	blockCreationTimeInSeconds := uint(blockCreationTime.Seconds())
+
+	expectedInclusionTime := uint(expectedInclusionInBlock) * blockCreationTimeInSeconds
+	if expectedInclusionTime < EstimatedTime5Sec {
+		return EstimatedTime5Sec
+	} else if expectedInclusionTime < EstimatedTime10Sec {
+		return EstimatedTime10Sec
+	} else if expectedInclusionTime < EstimatedTime15Sec {
+		return EstimatedTime15Sec
+	} else if expectedInclusionTime < EstimatedTime20Sec {
+		return EstimatedTime20Sec
+	} else if expectedInclusionTime < EstimatedTime30Sec {
+		return EstimatedTime30Sec
+	} else if expectedInclusionTime < EstimatedTime45Sec {
+		return EstimatedTime45Sec
+	} else if expectedInclusionTime < EstimatedTime1Min {
+		return EstimatedTime1Min
+	}
+
+	return EstimatedTimeMoreThan1Min
+}
+
+func (f *FeeManager) estimatedTimeV2(feeHistory *FeeHistory, txMaxFeePerGas *big.Int, txPriorityFee *big.Int, chainID uint64) uint {
+	sortedBaseFees := f.convertToBigIntAndSort(feeHistory.BaseFeePerGas)
+	if len(sortedBaseFees) == 0 {
+		return EstimatedTimeUnknown
+	}
+
+	var mediumPriorityFees []string // based on 50th percentile in the last 100 blocks
+	for _, fee := range feeHistory.Reward {
+		mediumPriorityFees = append(mediumPriorityFees, fee[0])
+	}
+	mediumPriorityFeesSorted := f.convertToBigIntAndSort(mediumPriorityFees)
+	if len(mediumPriorityFeesSorted) == 0 {
+		return EstimatedTimeUnknown
+	}
+
+	txBaseFee := new(big.Int).Sub(txMaxFeePerGas, txPriorityFee)
+
+	networkCongestion := calculateNetworkCongestion(feeHistory)
+
+	// Priority fee for the first two blocks has to be higher than 60th percentile of the mediumPriorityFeesSorted
+	priorityFeePercentileIndex := int(float64(len(mediumPriorityFeesSorted)) * 0.6)
```

Comment:
where are those numbers coming from? they seem a bit random to me.

Previous algorithm used mathematical probability which seems more robust to me, especially if those numbers are random

Reply:
Yes, looking just at the numbers, they look randomly selected, but they represent the 60th percentile of the medium priority fees sorted, which represents the 50th percentile of priority fees in the last 100 blocks.

> Previous algorithm used mathematical probability which seems more robust to me

Yes, that's true, but it didn't take the priority fee into account and also the `pEvent` was calculated for the current block and the possibility of inclusion in the next `X` blocks was calculated as the base fee will always be in the same range, while the min-max base fee range varies.

Reply:
@vkjr sure, done.